src/sentry/api/endpoints/organization_dashboard_details.py

```diff
@@ -1,11 +1,97 @@
+from rest_framework.response import Response
+
 from sentry.api.base import DocSection
 from sentry.api.bases.dashboard import (
     OrganizationDashboardEndpoint
 )
 from sentry.api.serializers import serialize
-from sentry.models import ObjectStatus
+from sentry.api.serializers.rest_framework import ListField, ValidationError
+from sentry.models import ObjectStatus, Widget
+
+
+def get_next_dashboard_order(dashboard_id):
+    max_order = Widget.objects.filter(
+        dashboard_id=dashboard_id,
+    ).aggregate(Max('order'))
+    return max_order['order__max'] + 1
+
+
+def remove_widgets(dashboard_widgets, widget_data):
+    """
+    Removes current widgets belonging to dashboard not in widget_data.
+    Returns remaining widgets.
+    """
+    widget_titles = [wd['id'] for wd in widget_data]
+    dashboard_widgets.exclude(
+        id__in=widget_titles
+    ).update(status=ObjectStatus.PENDING_DELETION)
+
+    return dashboard_widgets.filter(
+        id__in=widget_titles
+    )
+
+
+def reorder_widgets(dashboard_id, widget_data):
+    """
+    Reorders Widgets given the relative order desired,
+    reorders widgets in the next possible set of numbers
+    i.e if order of widgets is 1, 2, 3
+        the reordered widgets will have order 4, 5, 6
+    """
+    dashboard_widgets = Widget.objects.filter(
+        dashboard_id=dashboard_id,
+    )
+    dashboard_widgets = list(remove_widgets(dashboard_widgets, widget_data))
+
+    # dashboard_widgets and widget_data should now have the same widgets
+    widget_data.sort(key=lambda x: x['order'])
+
+    next_order = get_next_dashboard_order(dashboard_id)
+    for index, data in enumerate(widget_data):
+        for widget in dashboard_widgets:
+            if widget.id == data['id']:
+                widget.order = next_order + index
+                widget.save()
+                break
+
+
+class WidgetSerializer(serializers.Serializer):
+    order = serializers.IntegerField(min_value=0, required=True)
+    id = serializers.IntegerField(min_value=0, required=True)
+
+
+class DashboardWithWidgetsSerializer(serializers.Serializer):
+    title = serializers.CharField(required=False)
+    widgets = ListField(
+        child=WidgetSerializer(),
+        required=False,
+        allow_null=True,
+    )
+
+    def validate_widgets(self, attrs, source):
+        try:
+            widgets = attrs[source]
+        except KeyError:
+            return attrs
+
+        if len(widgets) != len(set([w['order'] for w in widgets])):
+            raise ValidationError('Widgets must have no repeating order')
+
+        widgets_count = len(Widget.objects.filter(
+            id__in=[w['id'] for w in widgets],
+            dashboard_id=self.context['dashboard_id'],
+            status=ObjectStatus.VISIBLE,
+        ))
+
+        if len(widgets) != widgets_count:
+            raise ValidationError(
+                'All widgets must exist within this dashboard prior to reordering.')
```

Comment:
How does one add a widget?

Reply:
@markstory I think I mentioned in the pr description that this is only for reordering widgets. That right now is the only bulk editing needed. But adding and editing individual widgets are handled by the widgets endpoint which is in another PR